python/distant_vfx/colour/utilities.py

```diff
@@ -0,0 +1,189 @@
+    glob_name_template = "{shot}_{descriptor}_v{version}".format(
+        shot=version_shot,
+        descriptor="*",
+        version="*"
+    )
+
+    scan_folder = os.path.join(
+        post_folder, "scan"
+    )
+
+    glob_path = os.path.join(
+        scan_folder,
+        glob_name_template
+    )
+
+    scans = {}
+    for matched_path in glob.iglob(glob_path):
+        data_struct = descriptor_from_filename(matched_path)
+        descriptor = data_struct.get("descriptor")
+        version = data_struct.get("version")
+        if descriptor not in scans:
+            scans[descriptor] = {}
+
+        if version not in scans[descriptor]:
+            scans[descriptor][version] = {}
+
+        scans[descriptor][version] = data_struct
+
+    if "bg" in scans:
+        # Use the bg first
+        cube_path = search_for_cube_by_descriptor(
+            descriptor="bg",
+            plates=scans,
+            root_dir=scan_folder
+        )
+        if cube_path:
+            return cube_path
+    for descriptor in scans:
+        cube_path = search_for_cube_by_descriptor(
+            descriptor=descriptor,
+            plates=scans,
+            root_dir=scan_folder
+        )
+        if cube_path:
+            return cube_path
+    return None
+
+
+def _get_proj_seq_shot_descriptor_version(path):
+    basename = os.path.basename(path).replace(".", "_")
+    splits = basename.split("_")
+    version_splits = [
+        split[1:]
+        for split in splits
+        if split.startswith("v") and split[1:].isdigit()
+    ]
+
+    return (
+        "dst",
+        basename[:3].lower(),
+        basename[:7].lower(),
+        splits[1].lower() if len(splits) > 1 else "",
+        version_splits[0] if len(version_splits) else "001"
+    )
+
+
+def descriptor_from_filename(filename):
+    project, sequence, shot, descriptor, version = _get_proj_seq_shot_descriptor_version(filename)
+
+    try:
+        number_version = int(version)
+    except:
+        number_version = 1
+    data_struct = {
+        "project": project,
+        "sequence": sequence,
+        "shot": shot,
+        "descriptor": descriptor,
+        "version": number_version,
+        "filename": filename,
+    }
+
+    return data_struct
+
+
+def search_for_cube_by_descriptor(descriptor, plates, root_dir):
+    # Use the bg first
+    background_versions = plates.get(descriptor, [])
+    for version_num in sorted(background_versions.keys(), reverse=True):
+        data_struct = background_versions[version_num]
+
+        cube_path = os.path.join(
+            root_dir,
+            data_struct.get("filename"),
+            "_cdl",
+            "versioned",
+            data_struct.get("filename"),
+            "luts",
+            data_struct.get("filename") + ".cube"
+        )
+        if os.path.exists(cube_path):
+            # Perfect! Use it!
+            return cube_path
+    return None
+
+
+def _path_get_shot_lut_from_path(path):
```

Comment:
ShotLUT logic starts here....
Basically search Shot plates folder then if you can't find it there fall back to the scan folder.

We want to see if our render basename, eg `ftc1000_comp_v003.%04d.exr`, can resolve back to the latest cdl.
In this case an example would be `ftc1000_bg_v001.cube` or something.
Now we want to prefer the bg colour because that is most likely to be correct.

So we listdir the plates dir at   ../../../../plate (relative to the rendered frame).
We might get a list like the following:
    ftc1000_bg_v001
    ftc1000_gb_v001
    ftc1000_sb_v001
    ftc1000_bg_v002

We group it like this:

{
    "bg": {
        1: {},
        2: {},
    },
    "sb": {
        1: {},
    },
    "gb": {
        1: {},
    },
}

We search a sorted list of the "bg", so we get "ftc1000_bg_v002" first, looking in there, we see if we can find a cube file within there.
If not, we search the "ftc1000_bg_v001" folder and then latest the sb and gb versions.


If we still haven't found it we want to revert to the "scan" dir. So we go to `_find_cube_path_from_scan_dir` and using a wildcard glob search for "<vfxID>_*_v*" so basically "ftc1000_*_v*" and run the same logic as if the returned folders were a "plate" folder in a shot.

Reply:
This makes sense to me, I read through the code and nothing jumped out at me. One question, I might have missed this, where does the sorting happen to grab the latest bg plate?

Reply:
```

def search_for_cube_by_descriptor(descriptor, plates, root_dir):
    # Use the bg first
    background_versions = plates.get(descriptor, [])
    for version_num in sorted(background_versions.keys(), reverse=True):
        data_struct = background_versions[version_num]

```

Reply:
Ahhh gotcha, sorry I missed it!